scipy/optimize/_numdiff.py

```diff
@@ -111,14 +113,70 @@ def _prepare_bounds(bounds, x0):
     return lb, ub
 
 
+def group_columns(A, order=None):
+    """Group columns of a 2-d matrix for sparse finite differencing [1]_.
+
+    Two columns are in the same group if in each row at least one of them
+    has zero. A greedy sequential algorithm is used to construct groups.
+
+    Parameters
+    ----------
+    A : array_like or sparse matrix, shape (m, n)
+        Matrix of which to group columns.
+    order : None or iterable of int, shape (n,)
+        Permutation array which defines the order of columns enumeration.
+        If None (default) the order will be random.
+
+    Returns
+    -------
+    groups : ndarray of int, shape (n,)
+        Contains values from 0 to n_groups-1, where n_groups is the number
+        of found groups. Each value ``groups[i]`` is an index of a group to
+        which i-th column assigned. The procedure was helpful only if
+        n_groups is significantly less than n.
+
+    References
+    ----------
+    .. [1] A. Curtis, M. J. D. Powell, and J. Reid, "On the estimation of
+           sparse Jacobian matrices", Journal of the Institute of Mathematics
+           and its Applications, 13 (1974), pp. 117-120.
+    """
+    if issparse(A):
+        A = csc_matrix(A)
+    else:
+        A = np.atleast_2d(A)
+        A = (A != 0).astype(np.int32)
+
+    if A.ndim != 2:
+        raise ValueError("`A` must be 2-dimensional.")
+
+    m, n = A.shape
+
+    if order is None:
+        order = np.arange(n)
+        np.random.shuffle(order)
```

Comment:
Totally agree, but it won't be used anywhere. I will add anyway (maybe for later).